Add alias support for sparse tensors.

pearu opened the pull request:
This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104145
* __->__ #103810
* #108512

pearu committed:
Add alias support for sparse tensors.

[ghstack-poisoned]

pearu committed:
Update on "Add alias support for sparse tensors."

This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).




[ghstack-poisoned]

pearu committed:
Update on "Add alias support for sparse tensors."

This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3651,6 +3651,17 @@ Tensor view(const Tensor& self,
 }
 
 Tensor alias(const Tensor& self) {
+  switch (self.layout()) {
```
Why not use `native_functions.yaml` specific keys to call these ops?

pearu replied:
Fixed.

albanD commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -529,11 +529,15 @@ SparseTensor dense_to_sparse(const Tensor& self, c10::optional<c10::Layout> layo
+        // RuntimeError: result.storage().use_count() == 1 INTERNAL ASSERT FAILED
+        return self.clone();
+      } else {
+        return self.alias();
```
alias() here is technically not ok as the signature of this function says the output is not a view of the input (that's why the Dense case fails).
This can lead to silent correctness issues with autograd if you care about it.

pearu replied:
This code bit is out-of-date. So, resolving the discussion.

albanD reviewed (commented):
Change sounds ok. We need some testing.

pearu committed:
Update on "Add alias support for sparse tensors."

This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).




[ghstack-poisoned]

pearu committed:
Update on "Add alias support for sparse tensors."

This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).




[ghstack-poisoned]

pearu committed:
Update on "Add alias support for sparse tensors."

This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).




[ghstack-poisoned]

albanD reviewed (commented):
Sounds good!
Are you planning on adding test for this?

pearu commented:
> Sounds good! Are you planning on adding test for this?

I wasn't sure how to test `alias` from Python (`torch.alias` does not exist although `torch.alias_copy.__doc__` references it) until I found `torch.ops.aten.alias`...

@albanD , do we have a specification for `alias`? In order words, what are the basic properties of `alias` that the tests should cover?

alexsamardzic commented:
As the PR seems to be rebased on a branch that includes [PR 92022](https://github.com/pytorch/pytorch/pull/92022), would it be good to include the change you suggested in [this comment](https://github.com/pytorch/pytorch/pull/92022#discussion_r1234909703) in this PR too?

albanD commented:
@pearu yes it is a bit of a weird one. If we don't have an opinfo for `torch.ops.aten.alias()` then I would suggest a small on-off test for sparse to make sure it behaves as expect (returns a view as expected).

cpuhrsch reviewed (changes requested):
Still needs tests

pearu committed:
Update on "Add alias support for sparse tensors."

This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).




[ghstack-poisoned]

pearu committed:
Update on "Add alias support for sparse tensors."

This PR addresses the sparse tensors part of https://github.com/pytorch/pytorch/issues/99655

The PR introduces the following utility functions:
- `at::sparse_csr::alias_with_values(a_sparse_compressed_tensor, new_values)`
- `at::sparse::alias_with_values(a_sparse_tensor, new_values)`

These functions return a wrapper of a sparse tensor with new specified values that allow introducing alias support for sparse tensors and more (e.g. the most efficient way to resolve https://github.com/pytorch/pytorch/pull/99292#discussion_r1232201104 is to use `at::sparse_csr::alias_with_values(self, self.values().to(dtype))` as a replacement of `self.to(dtype)` to avoid the unnecessary copy of indices).




[ghstack-poisoned]

pearu commented:
Converting to draft as there is no immediate need for the feature.
